File: app/src/main/java/com/sevtinge/cemiuiler/module/securitycenter/lab/BlurLocationEnable.java
package com.sevtinge.cemiuiler.module.securitycenter.lab;

import android.content.ComponentName;
import com.sevtinge.cemiuiler.module.base.BaseHook;
import com.sevtinge.cemiuiler.module.securitycenter.SecurityCenterDexKit;
import com.sevtinge.cemiuiler.utils.Helpers;
import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.XposedBridge;
import de.robv.android.xposed.XposedHelpers;
import de.robv.android.xposed.callbacks.XC_LoadPackage;
import de.robv.android.xposed.callbacks.XCallback;
import io.luckypray.dexkit.descriptor.member.DexClassDescriptor;

import java.io.File;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import static com.sevtinge.cemiuiler.utils.Helpers.getPackageVersionCode;

public class BlurLocationEnable extends BaseHook {

    Class<?> mLab;
    Class<?> mStableVer;

    Class<?> utilCls;
    Class<?> labUtils;

    @Override
    public void init() {
        try {
            List<DexClassDescriptor> result = Objects.requireNonNull(SecurityCenterDexKit.mSecurityCenterResultClassMap.get("LabUtils"));
            for (DexClassDescriptor descriptor : result) {
                labUtils = descriptor.getClassInstance(lpparam.classLoader);
                log("labUtils class is " + labUtils);
                findAndHookMethod("com.miui.permcenter.settings.PrivacyLabActivity", "onCreateFragment", new MethodHook() {
                    @Override
                    protected void before(MethodHookParam param) throws Throwable {
                        Object fm = Helpers.getStaticObjectFieldSilently(labUtils, "b");
                        if (fm != null) {
                            try {
                                Map<String, Integer> featMap = (Map<String, Integer>) fm;
                                featMap.put("mi_lab_blur_location_enable", 0);
                                //featMap.put("mi_lab_blur_location_enable", 0);
                            } catch (Throwable ignore) {
                            }
                        }
                    }
                });
            }
        } catch (Throwable e) {
            e.printStackTrace();
        }
        /*mLab = findClassIfExists("com.miui.permcenter.q");
        mStableVer = findClassIfExists("miui.os.Build");

        findAndHookMethod(mLab, "h", new MethodHook() {
            @Override
            protected void before(MethodHookParam param) throws Throwable {
                param.setResult(true);
            }
        });*/

        /*
        Helpers.findAndHookMethod("com.miui.permcenter.settings.PrivacyLabActivity", lpparam.classLoader, "onCreateFragment", new MethodHook() {
            @Override
            protected void before(MethodHookParam param) throws Throwable {
                int appVersionCode = getPackageVersionCode(lpparam);
                if (appVersionCode >= 40000749) {
                    utilCls = findClassIfExists("rb.e", lpparam.classLoader);
                } else {
                    utilCls = findClassIfExists("com.miui.permcenter.utils.h", lpparam.classLoader);
                }
                if (utilCls != null) {
                    Object fm = Helpers.getStaticObjectFieldSilently(utilCls, "b");
                    if (fm != null) {
                        try {
                            Map<String, Integer> featMap = (Map<String, Integer>) fm;
                            //featMap.put("mi_lab_ai_clipboard_enable", 0);
                            featMap.put("mi_lab_blur_location_enable", 0);
                        } catch (Throwable ignore) {
                        }
                    }
                }
            }
        });

         */

        //findAndHookMethod(mStableVer, "IS_STABLE_VERSION", new MethodHook() {
        //    @Override
        //    protected void before(MethodHookParam param) throws Throwable {
        //        param.setResult(true);
        //    }
        //});
    }

}

File: app/src/main/java/com/sevtinge/hyperceiler/module/hook/securitycenter/lab/BlurLocationEnable.kt
/*
  * This file is part of HyperCeiler.

  * HyperCeiler is free software: you can redistribute it and/or modify
  * it under the terms of the GNU Affero General Public License as
  * published by the Free Software Foundation, either version 3 of the
  * License.

  * This program is distributed in the hope that it will be useful,
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  * GNU Affero General Public License for more details.

  * You should have received a copy of the GNU Affero General Public License
  * along with this program.  If not, see <https://www.gnu.org/licenses/>.

  * Copyright (C) 2023-2024 HyperCeiler Contributions
*/
package com.sevtinge.hyperceiler.module.hook.securitycenter.lab

import com.sevtinge.hyperceiler.module.base.*
import com.sevtinge.hyperceiler.module.hook.securitycenter.lab.LabUtilsClass.labUtilClass


object BlurLocationEnable : BaseHook() {
    private var labUtils: Class<*>? = null

    override fun init() {
        labUtilClass.forEach {
            labUtils = it
            logI(TAG, this.lpparam.packageName, "labUtils class is $labUtils")
            findAndHookMethod(
                "com.miui.permcenter.settings.PrivacyLabActivity",
                "onCreateFragment",
                object : MethodHook() {
                    @Throws(Throwable::class)
                    override fun before(param: MethodHookParam) {
                        val fm = getStaticObjectFieldSilently(labUtils, "b")
                        if (fm != null) {
                            try {
                                val featMap = fm as MutableMap<String, Int>
                                featMap["mi_lab_blur_location_enable"] = 0
                                // featMap.put("mi_lab_blur_location_enable", 0);
                            } catch (ignore: Throwable) {
                            }
                        }
                    }
                })
        }

        /*try {
            val result: List<DexClassDescriptor> = Objects.requireNonNull<List<DexClassDescriptor>>(
                SecurityCenterDexKit.mSecurityCenterResultClassMap.get("LabUtils")
            )
            for (descriptor in result) {
                labUtils = descriptor.getClassInstance(lpparam.classLoader)
                log("labUtils class is $labUtils")
                findAndHookMethod(
                    "com.miui.permcenter.settings.PrivacyLabActivity",
                    "onCreateFragment",
                    object : BaseHook.MethodHook() {
                        @Throws(Throwable::class)
                        protected override fun before(param: MethodHookParam) {
                            val fm = Helpers.getStaticObjectFieldSilently(labUtils, "b")
                            if (fm != null) {
                                try {
                                    val featMap = fm as MutableMap<String, Int>
                                    featMap["mi_lab_blur_location_enable"] = 0
                                    // featMap.put("mi_lab_blur_location_enable", 0);
                                } catch (ignore: Throwable) {
                                }
                            }
                        }
                    })
            }
        } catch (e: Throwable) {
            e.printStackTrace()
        }*/
    }
}
